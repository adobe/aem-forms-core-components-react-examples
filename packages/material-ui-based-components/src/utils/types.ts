import { FieldJson, FieldsetJson, State } from '@aemforms/af-core';
import { Handlers, WithViewState } from '@aemforms/af-react-renderer';

export type FieldViewState = WithViewState<FieldJson>;

export type PROPS = State<FieldJson & Handlers & {
  isError?: boolean,
  layout?: {
    [key: string]: any
  }
  text?:string,
  richText?:boolean
}>;

export type PROPS_PANEL = State<FieldsetJson> & { handlers: Handlers }
