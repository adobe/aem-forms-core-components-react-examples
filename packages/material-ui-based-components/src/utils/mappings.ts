import Form from '../components/Form';
import Panel from '../components/Panel';
import TextField from '../components/TextField';
import Date from '../components/Date';
import PlainText from '../components/PlainText';
import Title from '../components/Title';
import Image from '../components/Image';
import Rating from '../components/Rating';
import Slider from '../components/Slider';
import TabsVerticalComp from '../components/TabsVerticalComp';
import TabsHorizontalComp from '../components/TabsHorizontalComp';

export default {
  form: Form,
  panel: Panel,
  'text-input': TextField,
  'date-input': Date,
  'text-input': TextField,
  'plain-text': PlainText,
  'custom:title': Title,
  image: Image,
  'text-input': TextField,
  'custom:rating':Rating,
  'custom:slider':Slider
  'text-input': TextField,
  'custom:vertical-tab': TabsVerticalComp,
  'custom:horizontal-tab': TabsHorizontalComp
};