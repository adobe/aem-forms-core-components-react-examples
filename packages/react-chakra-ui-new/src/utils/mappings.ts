// *******************************************************************************
//  * Copyright 2023 Adobe
//  *
//  * Licensed under the Apache License, Version 2.0 (the “License”);
//  * you may not use this file except in compliance with the License.
//  * You may obtain a copy of the License at
//  *
//  *     http://www.apache.org/licenses/LICENSE-2.0
//  *
//  * Unless required by applicable law or agreed to in writing, software
//  * distributed under the License is distributed on an “AS IS” BASIS,
//  * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
//  * See the License for the specific language governing permissions and
//  * limitations under the License.
//  ******************************************************************************

import Button from '../components/Button';
import CheckBox from '../components/CheckBox';
import CheckBoxGroup from '../components/CheckBoxGroup';
import DropDown from '../components/DropDown';
import Number from '../components/Number';
import RadioGroup from '../components/RadioButton';
import TextField from '../components/TextField';
import Panel from '../components/Panel';
import Wizard from '../components/Wizard';
import Accordion from '../components/Accordion';
import HorizontalTab from '../components/HorizontalTabs';
import VerticalTab from '../components/VerticalTabs';
import Form from '../components/Form';
import Image from '../components/Image';
import PlainText from '../components/PlainText';

const mappings = {
  'text-input': TextField,
  'password-input': TextField,
  'multiline-input': TextField,
  panel: Panel,
  'checkbox-group': CheckBoxGroup,
  'radio-group': RadioGroup,
  'drop-down': DropDown,
  'number-input': Number,
  button: Button,
  checkbox: CheckBox,
  image: Image,
  'core/fd/components/form/wizard/v1/wizard': Wizard,
  'core/fd/components/form/accordion/v1/accordion': Accordion,
  'core/fd/components/form/tabsontop/v1/tabsontop': HorizontalTab,
  'core/fd/components/form/verticaltabs/v1/verticaltabs': VerticalTab,
  'plain-text': PlainText,
  'core/fd/components/form/panelcontainer/v1/panelcontainer': Panel,
  form: Form
};

export default mappings;