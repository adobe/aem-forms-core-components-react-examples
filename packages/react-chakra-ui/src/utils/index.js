import React from 'react';
import { FormContext } from '@aemforms/af-react-renderer';
import { createFormInstance } from '@aemforms/af-core';
import { ChakraProvider } from '@chakra-ui/react'
import { IntlProvider } from 'react-intl';
import { render } from '@testing-library/react';

export const DEFAULT_ERROR_MESSAGE = 'There is an error in the field';

export const createForm = (field) => {
  const formJson = {
    items: [field]
  };
  return createFormInstance(formJson);
};

export const Provider = (form, mappings, locale, dictionaries) => props => {
  const c = {
    form,
    mappings,
    modelId: '$form',
    refMap: {}
  };
  const { children } = props;
  return (
    <ChakraProvider>
      <IntlProvider locale={locale} messages={dictionaries}>
        <FormContext.Provider value={c}>{children}</FormContext.Provider>
      </IntlProvider>
    </ChakraProvider>
  );
};

export const renderComponent = function (Component) {
  const test = (field, operation) => {
    const form = createForm(field);
    if (operation) {
      operation(form, form.items[0]);
    }
    const e = form.items[0].getState();
    let component = <Component {...e} />;
    const wrapper = Provider(form);
    const renderResponse = render(component, { wrapper });
    return {
      renderResponse,
      form,
      element: form?.items[0],
      component
    };
  };
  return test;
};
