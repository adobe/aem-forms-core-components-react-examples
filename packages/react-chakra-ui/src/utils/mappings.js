import TextField from "../generic_components/TextFieldComp";
import DateField from "../components/DateFieldComp";
import ButtonComp from "../generic_components/ButtonComp";
import CheckBoxComp from "../generic_components/CheckBoxComp";
import TextAreaComp from "../generic_components/TextAreaComp";
import Panel from "../generic_components/Panel";
import CustomPanel from "../components/CustomPanel";
import CustomText from "../components/CustomText";
import CustomTextButton from "../components/CustomTextButton";
import CustomHeading from "../components/CustomHeading";
import SelectComp from "../generic_components/SelectComp";
import VerticalContainer from "../components/VerticalContainer";
import HorizontalWrap from "../components/HorizontalWrap";
import CheckBoxGroup from "../components/CheckBoxGroup";
import TabsComp from "../components/TabsComp";
import RadioComp from "../generic_components/RadioComp";
import AccordionComp from "../generic_components/AccordionComp";
import LoanButtonComp from "../components/LoanButtonComp"; 
import CustomTextAreaComp from "../components/CustomTextArea";
import Form from "../components/Form";

const myMapping = {
  "text-input": TextField,
  "date-input": DateField,
  "number-input": TextField,
  "multiline-input": TextAreaComp,
  checkbox: CheckBoxComp,
  "checkbox-group": CheckBoxGroup,
  button: ButtonComp,
  panel: Panel,
  "custom:panel": CustomPanel,
  "custom:loanButton": LoanButtonComp,
  "drop-down": SelectComp,
  "radio-group": RadioComp,
  "custom:static/text": CustomText,
  "custom:static/textButton": CustomTextButton,
  "custom:static/heading": CustomHeading,
  "custom:vertical-flex": VerticalContainer,
  "custom:horizontal-flex": HorizontalWrap,
  "custom:vertical-tab": TabsComp,
  "custom:accordion": AccordionComp,
  "custom:accordionItem": CustomTextAreaComp,
  form: Form
};
export default myMapping;
