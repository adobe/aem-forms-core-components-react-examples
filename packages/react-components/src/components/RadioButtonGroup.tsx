import React from 'react';
import { useRuleEngine } from '@aemforms/af-react-renderer';

const RadioButtonGroup = (fieldjson: any) => {
    const [props, handlers] = useRuleEngine(fieldjson);
    const { id, label, required, enumNames, enum: enums, name, valid, errorMessage, description } = props;
    const options = enumNames && enumNames.length ? enumNames : enums || [];
    const validateState = valid === false ? 'invalid' : 'valid';
    const error = validateState === 'invalid';
   
    const changeHandler = (event: React.ChangeEvent<HTMLInputElement>) => {
       const val = event.target.value;
       console.log('rad-val',val);
       handlers.dispatchChange(val);
     };
    return (
       <div>
         <div className="cmp-adaptiveform-radiobutton">
             <label htmlFor={id} className="cmp-adaptiveform-radiobutton__label">{label?.value}</label>
         <div className="cmp-adaptiveform-radiobutton__widget HORIZONTAL">
             {options?.map((item: string, index: number) => {
                 return (
                     <div className="cmp-adaptiveform-radiobutton-item [${id}]">
                       <input className="cmp-adaptiveform-radiobutton__widget" type='radio' required={required} name={name} value={enumNames[index]} onChange={changeHandler} />
                       <label className="cmp-adaptiveform-radiobutton__label"><span>{item}</span></label>
                     </div>
                 );
             })};
         </div>
          </div>
          {error && <div className="cmp-adaptiveform-radiobutton__errormessage">{errorMessage}</div>}
          {description && !error && <div className="cmp-adaptiveform-radiobutton__shortdescription">{description}</div>}
       </div>
    );
 };

export default RadioButtonGroup;