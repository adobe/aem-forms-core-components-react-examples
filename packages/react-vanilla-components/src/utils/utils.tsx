export const formatBytes = (bytes: number, decimals = 0) => {
    if (!+bytes) {
      return '0 Bytes';
    }
    const k = 1024;
    const dm = decimals < 0 ? 0 : decimals;
    const sizes = ['bytes', 'kb', 'mb', 'gb', 'tb', 'pb'];
    const i = Math.floor(Math.log(bytes) / Math.log(k));

    return `${parseFloat((bytes / Math.pow(k, i)).toFixed(dm))} ${sizes[i]}`;
  };

  export const syncAriaDescribedBy = (id: string, tooltip?: string, description?: string, error?: string) => {
    const descriptions = [];

    if (description) {
        descriptions.push(`${id}__longdescription`);
    }
    if (tooltip) {
        descriptions.push(`${id}__shortdescription`);
    }
    if (error) {
        descriptions.push(`${id}__errormessage`);
    }

    return descriptions.join(' ');
};