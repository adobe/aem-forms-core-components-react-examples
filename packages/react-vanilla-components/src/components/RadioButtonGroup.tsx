// *******************************************************************************
//  * Copyright 2023 Adobe
//  *
//  * Licensed under the Apache License, Version 2.0 (the “License”);
//  * you may not use this file except in compliance with the License.
//  * You may obtain a copy of the License at
//  *
//  *     http://www.apache.org/licenses/LICENSE-2.0
//  *
//  * Unless required by applicable law or agreed to in writing, software
//  * distributed under the License is distributed on an “AS IS” BASIS,
//  * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
//  * See the License for the specific language governing permissions and
//  * limitations under the License.

//  * The BEM markup is as per the AEM core form components guidelines.
//  * LINK- https://github.com/adobe/aem-core-forms-components/blob/master/ui.af.apps/src/main/content/jcr_root/apps/core/fd/components/form/radiobutton/v1/radiobutton/radiobutton.html
//  ******************************************************************************

import React, { useState } from 'react';
import { withRuleEngine } from '../utils/withRuleEngine';
import { PROPS } from '../utils/type';

const RadioButtonGroup = (props: PROPS) => {
  const { id, label, required, enumNames, enum: enums, name, isError, errorMessage, description, visible, enabled, value, readOnly } = props;
  const options = enumNames && enumNames.length ? enumNames : enums || [];
  const orientation = props.layout?.orientation.toUpperCase();
  const [shortDescription, setShortDescription] = useState(true);
  const [longDescription, setLongtDescription] = useState(false);

  const handleClick = (event: React.MouseEvent<HTMLButtonElement, MouseEvent>) => {
    event.preventDefault();
    setShortDescription(!shortDescription);
    setLongtDescription(!longDescription);
  };

  const changeHandler = (event: React.ChangeEvent<HTMLInputElement>) => {
    const val = event.target.value;
    props.dispatchChange(val);
  };
  return (
    <div className="cmp-adaptiveform-radiobutton" data-cmp-is="adaptiveFormRadioButton" data-cmp-visible={visible} data-cmp-enabled={enabled}>
      {label?.visible && <label htmlFor={`${id}-widget`} className="cmp-adaptiveform-radiobutton__label">{label?.value}</label>}
      {description && <button aria-label='Toggle Button' className="cmp-adaptiveform-radiobutton__questionmark" onClick={handleClick}></button>}
      <div className={`cmp-adaptiveform-radiobutton__widget ${orientation}`} id={`${id}-widget`}>
        {options?.map((item: string, index: number) => {
          return (
            <div className="cmp-adaptiveform-radiobutton__option" key={item}>
              <label className="cmp-adaptiveform-radiobutton__option__label">
                <input className={value === enums![index] ? 'cmp-adaptiveform-radiobutton__option__widget cmp-adaptiveform-radiobutton__option__widget--filled' : 'cmp-adaptiveform-radiobutton__option__widget cmp-adaptiveform-radiobutton__option__widget--empty'} type='radio' required={required} name={name} value={enums![index]} onChange={changeHandler} readOnly={readOnly} aria-checked={value === enums![index] ? 'true' : 'false'} />
                <span>{item}</span>
              </label>
            </div>
          );
        })}
      </div>
      {shortDescription && props?.tooltip && <div title='Help Text' data-cmp-visible={shortDescription} className='cmp-adaptiveform-radiobutton__shortdescription'>{props?.tooltip}</div>}
      <div aria-live="polite">
        {longDescription && description && !errorMessage ? <div title='Help Text' data-cmp-visible={longDescription} className="cmp-adaptiveform-radiobutton__longdescription">{description}</div> : null}
      </div>
      {isError ? <div className="cmp-adaptiveform-radiobutton__errormessage">{errorMessage}</div> : null}
    </div>
  );
};

export default withRuleEngine(RadioButtonGroup);