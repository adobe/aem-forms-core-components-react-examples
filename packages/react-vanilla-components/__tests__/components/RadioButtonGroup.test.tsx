/*
 * Copyright 2023 Adobe, Inc.
 *
 * Your access and use of this software is governed by the Adobe Customer Feedback Program Terms and Conditions or other Beta License Agreement signed by your employer and Adobe, Inc.. This software is NOT open source and may not be used without one of the foregoing licenses. Even with a foregoing license, your access and use of this file is limited to the earlier of (a) 180 days, (b) general availability of the product(s) which utilize this software (i.e. AEM Forms), (c) January 1, 2023, (d) Adobe providing notice to you that you may no longer use the software or that your beta trial has otherwise ended.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL ADOBE NOR ITS THIRD PARTY PROVIDERS AND PARTNERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */

import userEvent from "@testing-library/user-event";
import RadioButtonGroup from '../../src/components/RadioButtonGroup';
import { renderComponent } from '../utils';
import "@testing-library/jest-dom/extend-expect"

const field = {
  name: 'radio',
  visible: true,
  label: {
    value: 'radio group'
  },
  fieldType: 'radio-group',
  enum: [1, 2, 3],
  enumNames: [{
    value: "radio 1"
  },
  {
    value: "radio 2"
  },
  {
    value: "radio 3"
  }
]
};

const  fieldTwo = {
  name: 'radio',
  visible: true,
  label: {
    value: 'radio group'
  },
  fieldType: 'radio-group',
  enum: [1, 2, 3],
  enumNames: ['radio 1', 'radio 2', 'radio 3']
};
const helper = renderComponent(RadioButtonGroup);

describe('radio Group', () => {

  test('option selected by user is set in the model', async () => {
    const f = {
      ...field,
    };
    const { renderResponse, element } = await helper(f);

    expect(element?.getState().value).toBeUndefined();

    userEvent.click(renderResponse.getByText(f.enumNames[0].value));
    expect(element?.value).toEqual(1);

    userEvent.click(renderResponse.getByText(f.enumNames[2].value));
    expect(element?.value).toEqual(3);
  });

  test('option selected by user is set in the model', async () => {
    const f = {
      ...fieldTwo,
    };
    const { renderResponse, element } = await helper(f);

    expect(element?.getState().value).toBeUndefined();

    userEvent.click(renderResponse.getByText(f.enumNames[0]));
    expect(element?.value).toEqual(1);

    userEvent.click(renderResponse.getByText(f.enumNames[2]));
    expect(element?.value).toEqual(3);
  });

  test('selection made by the user sets the value', async () => {
    const f = {
      ...field,
    };
    const { renderResponse, element } = await helper(f);

    userEvent.click(renderResponse.getByText(f.enumNames[0].value));
    expect(element.value).toEqual(1);

    userEvent.click(renderResponse.getByText(f.enumNames[1].value));
    expect(element.value).toEqual(2);

    userEvent.click(renderResponse.getByText(f.enumNames[0].value));
    expect(element.value).toEqual(1);
  });

  test('selection made by the user sets the value', async () => {
    const f = {
      ...fieldTwo,
    };
    const { renderResponse, element } = await helper(f);

    userEvent.click(renderResponse.getByText(f.enumNames[0]));
    expect(element.value).toEqual(1);

    userEvent.click(renderResponse.getByText(f.enumNames[1]));
    expect(element.value).toEqual(2);

    userEvent.click(renderResponse.getByText(f.enumNames[0]));
    expect(element.value).toEqual(1);
  });

  test('it should handle visible property', async () => {
    const f = {
      ...field,
      visible: false,
    };
    const { renderResponse } = await helper(f);
    expect(renderResponse.queryByText(f.enumNames[0].value)).toBeNull();
    expect(renderResponse.queryByText(f.enumNames[1].value)).toBeNull();
    expect(renderResponse.queryByText(f.enumNames[2].value)).toBeNull();
  });

  test('In case of both tooltip and description, tooltip should be visible and onclick of toggle button, description should be visible', async () => {
    const f = {
      ...field,
      tooltip: 'Short Description',
      description: 'Mandatory',
      properties: {
        "afs:layout": {
            tooltipVisible: true
        },
      }
    };
    const helper = renderComponent(RadioButtonGroup);
    const { renderResponse } = await helper(f);
    expect(renderResponse.getByText('Short Description')).not.toBeNull();
    const button = renderResponse.container.getElementsByClassName('cmp-adaptiveform-radiobutton__questionmark');
    userEvent.click(button[0]);
    expect(renderResponse.getByText('Mandatory')).not.toBeNull();
  });

  test('html in the label should be handled for non rich text', async () => {
    const f = {
      ...field,
      label: {
        value: '<p>title inside p tags</p>',
        richText: true,
        visible: true
      }
    }
    let { renderResponse } = await helper(f);
    expect(renderResponse.container.innerHTML).toContain('<p>title inside p tags</p>');
  });

  test('Enumnames should be handled for non rich text', async () => {
    const f = {
      ...field,
      enumNames: field.enumNames.map((x) => ({
        ...x,
        value:'<i>radio 1</i>',
        richText: true
      })),
    }
    let { renderResponse } = await helper(f);
    expect(renderResponse.container.innerHTML).toContain('<i>radio 1</i>');
  });
  
  test('Enumnames should be handled for non rich text', async () => {
    const f = {
      ...field,
      enumNames: field.enumNames.map((x) => ({
        ...x,
        value:'<i>radio 1</i>',
        richText: false
      })),
    }
    let { renderResponse } = await helper(f);
    expect(renderResponse.container.innerHTML).toContain('radio 1');
  });

  test('Aria-describedby should contain long des short desc if present otherwise it should be empty', async () => {
    const f= {
      ...field,
      id: 'radio-123',
      tooltip: "short description",
      description: "long description",
      properties: {
        "afs:layout": {
            tooltipVisible: true
        },
      }
    };
    const { renderResponse } = await helper(f);
    const input = renderResponse.container.getElementsByClassName("cmp-adaptiveform-radiobutton__widget");
    expect(input).toHaveLength(1);
    expect(input[0]).toHaveAttribute('aria-describedby', `${f.id}__longdescription ${f.id}__shortdescription`)
  });
});