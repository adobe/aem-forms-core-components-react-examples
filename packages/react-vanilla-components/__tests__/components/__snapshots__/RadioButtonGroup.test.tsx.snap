// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`radio Group Snapshot 1`] = `
<div>
  <div
    class="cmp-adaptiveform-radiobutton cmp-adaptiveform-radiobutton--empty "
    data-cmp-enabled="true"
    data-cmp-is="adaptiveFormRadioButton"
    data-cmp-visible="true"
  >
    <div
      class="cmp-adaptiveform-radiobutton__label-container"
    >
      <label
        class="cmp-adaptiveform-radiobutton__label"
        for="SzJ1tfY3VW6-widget"
      >
        radio group
      </label>
    </div>
    <div
      class="cmp-adaptiveform-radiobutton__widget HORIZONTAL"
      id="SzJ1tfY3VW6-widget"
    >
      <div
        class="cmp-adaptiveform-radiobutton__option"
      >
        <label
          class="cmp-adaptiveform-radiobutton__option-label"
        >
          <input
            aria-checked="false"
            class="cmp-adaptiveform-radiobutton__option__widget"
            name="radio"
            type="radio"
            value="1"
          />
          <span>
            radio 1
          </span>
        </label>
      </div>
      <div
        class="cmp-adaptiveform-radiobutton__option"
      >
        <label
          class="cmp-adaptiveform-radiobutton__option-label"
        >
          <input
            aria-checked="false"
            class="cmp-adaptiveform-radiobutton__option__widget"
            name="radio"
            type="radio"
            value="2"
          />
          <span>
            radio 2
          </span>
        </label>
      </div>
      <div
        class="cmp-adaptiveform-radiobutton__option"
      >
        <label
          class="cmp-adaptiveform-radiobutton__option-label"
        >
          <input
            aria-checked="false"
            class="cmp-adaptiveform-radiobutton__option__widget"
            name="radio"
            type="radio"
            value="3"
          />
          <span>
            radio 3
          </span>
        </label>
      </div>
    </div>
    <div
      aria-live="polite"
    />
  </div>
</div>
`;
